### PR TITLE
Help and test for generators

### PR DESCRIPTION
**Pull request by abadams**
Here's a stab at adding virtual help and test methods to the base generator class. Not used for much here, because the existing generator tests were merged over from the AOT tests, but this will make writing generator jit tests much easier.

I intend to use the help and test methods when writing components for a larger library.

The help and test are accessible from the command line via:

list registered generators:
my_generator -list 

print help text for a given generator:
my_generator -help foo

test a generator:
my_generator -test foo

### CODE REVIEW

**Comment by steven-johnson** [comment]
I'm concerned that the test() method isn't the best design for this; it pretty much only allows for JIT-based testing, and I really think we need/want a solution that will work well for AOT as well. My first thought would be to have a separate Test class for each Generator, which could be placed in a separate source file, e.g.

    foo_generator.cc  // class Foo : public Generator<Foo>
    foo_generator_test.cc // class FooTest : public GeneratorTest<Foo>

this wouldn't be *quite* as seamless as the inline test method, but would offer us a lot more flexibility in the long run in terms of allowing AOT testing.

**Comment by abadams** [comment]
I'm trying to support lighter-weight testing than that. I like tests that
are immediately next to the code that they test, and I also like
self-contained single-file reusable components. Most of our current tests
are actually testing the compiler rather than testing an algorithm, so they
don't exactly reflect what I'm trying to design for. Here's the question
I've been pondering:

When testing an imaging algorithm written in Halide, when is a JIT-based
test not good enough?

The cases I can think of are better covered by whole-app tests (e.g.
checking for leaks), or wouldn't be served by our aot tests anyway (e.g.
checking performance on a mobile device), or are actually tests that the
compiler isn't buggy.

File: src/Generator.h [hunk 1]
```diff
@@ -381,6 +381,17 @@ class GeneratorBase : public NamesInterface {
     /** Build and return a Halide::Func. All Generators must override this. */
     virtual Func build() = 0;
 
+    /** Run a Generator's test. Returns whether or not it succeeds. If
+     * unimplemented, it is assumed to fail, because untested code is
+     * probably broken. */
+    virtual bool test() {return false;}
+
+    /** Print out help test describing how to use this generator. */
```
Comment (kree-colemcalughlin) [inline src/Generator.h]: Should this read "help text" rather than "help test"?

**Comment by kree-colemcalughlin** [comment]
I think I agree with Andrew for basic correctness testing, and like the idea of testing a generator being as easy as implementing the test() method. Although it could make sense to provide a separate AOT testing framework as well that has a system for benchmarking generated functions as well as comparing results to external libraries.

**Comment by abestephensg** [comment]
Andrew mentioned this, but JIT testing doesn't work if the JIT is not available, e.g. on mobile.

If we standardize a scheme to express a test case (input, call, validation) such that it could be realized either for a JIT run or in an AOT setup, then we could wrap it with the appropriate test harness in these environments.

**Comment by dsharletg** [comment]
Seconding Steven/Abe's point, I've had problems with testing in environments where JIT is problematic. I think an AOT based testing framework will be necessary, and as long as that is the case, it would be nice to have a testing framework that covers as many bases as possible with fewer code paths.

**Comment by steven-johnson** [comment]
> wouldn't be served by our aot tests anyway (e.g. checking performance on a mobile device)

I don't understand this statement. AFAIK, AOT tests are the only way to check performance on iOS.

**Comment by steven-johnson** [comment]
> I'm trying to support lighter-weight testing than that. I like tests that
> are immediately next to the code that they test, and I also like
> self-contained single-file reusable components.

I'm sympathetic to these goals, but I think the long-term negatives will be outweighed here: if we add the simple test() method, we'll still end up needing to add a separate testing facility to enable AOT testing and benchmarking, for platforms that don't allow us to JIT (to wit: iOS, which is too significant to ignore). I think we'd be better off biting the bullet and having a testing approach that is a little less convenient, rather than having two separate testing methodologies competing with each other.

**Comment by abadams** [comment]
Still thinking about the whole issue, but to clarify the confusing
statement: Our existing generator aot testing framework in the public repo
is not useful for testing performance on iOS, and I worry that nothing we
do will be. Those sorts of tests seem to end up heavily integrated with the
development environment for the particular platform and structured
according to the local testing methodologies.

**Comment by kree-colemcalughlin** [comment]
I'm still not sure I see why JIT testing on a platform where that is available would not be good enough for correctness testing. If you can test that your generated function is producing correct results on x86, then it should also produce correct results on iOS. If it doesn't that is a problem.

Running performance benchmarks for various platforms is not encompassed by this paradigm. But, to echo Andrew's point, maybe those are better handled by existing (or specialized) unit testing frameworks specific to the target platform. Do we really want to be in the business of writing a unit testing/benchmarking framework that works across x86, iOS, Android, Cuda, OpenCL, etc....?

**Comment by abestephensg** [comment]
> Those sorts of tests seem to end up heavily integrated with the development environment for the particular platform and structured according to the local testing methodologies.

I don't think it is a goal of any of the Halide tests to fit into local methodologies- these tend to be highly tailored to specific production environments which are different at every company.  Halide tests provide a developer reasonable assurance that a Halide build works for JIT on certain platforms.  We have the opportunity to expand that capability to AOT platforms by standardizing a mechanism around Generator.

>  Do we really want to be in the business of writing a unit testing/benchmarking framework that works across x86, iOS, Android, Cuda, OpenCL, etc....?

Halide gives the developer the ability to write a Func in a Generator and schedule/realize it on all of those platforms.  The heavy lifting of launching filters on x86, CUDA, OpenGL, etc. is part of the runtime already- what's not there is a way to express and realize the test (the input, call to make, how to validate).

**Comment by abestephensg** [comment]
Can we standardize on the test() method- preserve the C++ written input initialization and output validation logic, but add constraints to enable it to be executed either at Generator runtime (in the JIT), or at app runtime in the AOT generated object? That's not too far away from the current PR:

Let's add the constraint that you have to be able to JIT the test() function-

To test at Generator runtime via the JIT, the module containing the test would be wired up to the module containing the Halide generated func. The test would call the generated Func as an ordinary function.  In the AOT case, the Generator runtime would have both modules and could put both into its output object file.  We would add thin wrapper apps to call the generated test() function on each platform.

Constraints on the test implementation would reflect the Halide runtime: it would live in a separate file from the Generator, no std library, etc. but additional functionality could be added to the Halide runtime if necessary.

**Comment by abadams** [comment]
I'm not crazy about the idea of having one file that's compiled in two very different ways. This sounds potentially quite confusing for the test writer.

AOT tests in a separate file don't need much help from us. Some of the stuff in apps/support (like the Image class that can cast itself to a buffer_t *) is pretty useful, but the intention behind the buffer_t-based interface is that any given project probably has its own Image class, and the owners of the project can write a shim that maps it to a buffer_t *.

I think we're in a situation where we can add a lot of convenience to JIT-based testing, but there's not much we can do to help AOT testing. So we're stalled while we try to figure out if we can standardize both under the one banner.

After thinking about it this week, I'm still leaning towards the original approach. Correctness tests are something that should be run at build time, where libHalide must be available anyway. A generator binary should own everything to do with a generator: the source code, the documentation, and the tests. It should be capable of (cross-)compiling the source, printing the help, and running the tests. I think asking it to cross-compile the tests too is not worth the complexity.

Unless you're debugging the Halide compiler, it doesn't matter where the correctness tests are run, so let's make it as convenient as we can. On-device performance testing should be handled via HL_PROFILE or similar. Other tests (does this pipeline play nice with my gl context) need to be done in the context of the deployed app.

**Comment by abadams** [comment]
After discussions in person, I'm going to drop this PR for now, unless we have any brighter ideas, AOT tests in a separate file (without any extra support) will be the way to test generators.

**Comment by kree-colemcalughlin** [comment]
Does it make sense to merge the help stuff without the JIT testing? It seems like there are some usefule features there.

**Comment by abadams** [comment]
Oh, good point. The automatic parameter listing might be useful.

**Comment by abadams** [comment]
Reopened so I don't forget about the help stuff.

**Comment by abadams** [comment]
This was mostly decided against and is pretty out-of-date at this point. Reclosing.

**Pull request closed without merging**